### PR TITLE
GFORMS-1162 new calendarDate component

### PR DESCRIPTION
**Pull request by girishkamat**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/validation/CalendarDateValidation.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      .andThen(validateDayMonthCombo(formComponent, _))
+
+  private def validateDayMonthAreNumbers(formComponent: FormComponent): ValidatedType[(Int, Int)] =
+    formComponent.multiValueId.atomsModelComponentIds.map(formModelVisibilityOptics.data.one) match {
+      case Some(day) :: Some(month) :: Nil =>
+        (validInt(formComponent, day, messages("date.day")), validInt(formComponent, month, messages("date.month")))
+          .mapN((day, month) => (day, month))
+      case _ =>
+        validationFailure(formComponent.firstAtomModelComponentId, formComponent, "date.isMissing", None, "")
+    }
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/validation/CalendarDateValidation.scala:50]: There is missing validation on number of characters entered. Both `day` and `month` accept unlimited number of leading `0`.

Would it be possible to reuse `hasMaximumLength` and `isNumeric`(and maybe others) from `DateValidationLogic`?

**Review by josef-vlach — CHANGES_REQUESTED** [review]

**Comment by josef-vlach** [comment]
Also click on error link doesn't focus first field of the component.

File: test/uk/gov/hmrc/gform/models/FormModelSupport.scala [hunk 2]
```diff
@@ -43,7 +43,7 @@ trait FormModelSupport extends GraphSpec {
 
   val retrievals: MaterialisedRetrievals = AnonymousRetrievals(SessionId("dummy-sessionId"))
   val thirdPartyData: ThirdPartyData = ThirdPartyData.empty
-  val envelopeId: EnvelopeId = EnvelopeId("dummy")
+  val someEnvelopeId: EnvelopeId = EnvelopeId("dummy")
```
[the remark this answers is not in the document]
Comment (girishkamat) [reply test/uk/gov/hmrc/gform/models/FormModelSupport.scala:46]: CalendarDateValidationSpec uses FormModelSupport and ExampleData, and both seem to define envelopeId. I have made a fix to revert the renaming

File: app/uk/gov/hmrc/gform/validation/CalendarDateValidation.scala (hunk 1 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/validation/CalendarDateValidation.scala]: Done

**Review by josef-vlach — APPROVED** [review]

**Comment by girishkamat** [comment]
> Also click on error link doesn't focus first field of the component.

Noticed while testing, when day-month combo was invalid, clicking on the link did not focus correctly. Fixed

**Pull request merged**